pythonforandroid/recipes/android/src/android/permissions.py

```diff
@@ -422,13 +423,97 @@ class Permission:
         )
 
 
-def request_permissions(permissions):
-    python_activity = autoclass('org.kivy.android.PythonActivity')
-    python_activity.requestPermissions(permissions)
+class _onRequestPermissionsCallback(PythonJavaClass):
+    """Callback class for registering a Python callback from
+    onRequestPermissionsResult in PythonActivity.
+    """
+    __javainterfaces__ = ['org.kivy.android.PythonActivity$PermissionsCallback']
+    __javacontext__ = 'app'
+
+    def __init__(self, func):
+        self.func = func
+        super().__init__()
+
+    @java_method('(I[Ljava/lang/String;[I)V')
+    def onRequestPermissionsResult(self, requestCode,
+                                   permissions, grantResults):
+        self.func(requestCode, permissions, grantResults)
+
+
+class _RequestPermissionsManager:
+    """Internal class for requesting Android permissions via
+    requestPermissions, including registering callbacks to requestPermissions.
+
+    Permissions are requested through the method 'request_permissions' which
+    accepts a list of permissions and an optional callback.
+
+    Any callback will asynchronously receive arguments from
+    onRequestPermissionsResult on PythonActivity after requestPermissions is
+    called.
+
+    The callback supplied must accept two arguments: 'permissions' and
+    'grantResults' (as supplied to onPermissionsCallbackResult).
+
+    Note that calling request_permission on SDK_INT < 23 will return
+    immediately (as run-time permissions are not required), and so the callback
+    will never happen. Therefore, request_permissions should only be called
+    with a callback if calling check_permission has indicated that it is
+    necessary.
+
+    The attribute '_java_callback' is initially None, but is set when the first
+    permissions request is made. It is set to an instance of
+    onRequestPermissionsCallback, which allows the Java callback to be
+    propagated to the class method 'python_callback'. This is then, in turn,
+    used to call an application callback if provided to request_permissions.
+    """
+    _java_callback = None
+    _callbacks = {1: None}
+    _callback_id = 1
+    _lock = threading.Lock()
```

Comment:
That one deserves a comment, just what you explained on the PR conversation is fine

Reply:
Actually I only put that in so that if multiple threads started asking for permissions at once, it wouldn't clobber the incrementing callback_id. But I can add a comment. I haven't tested it for thread safety; in my app I only ask for permissions from the main Kivy thread and not from other threads. If it's guaranteed that only one thread will _request_ permissions then it probably isn't needed.

Reply:
Well the code looks mergeable to me anyway, that was just a suggestion.
I did understand what you explained initially and yes it makes sense, once you explained it :smile:  I'm just thinking of people who will maintain that in 2 years and may wonder what it's for.
Usually we put code comments for things that are not so common/obvious like this. But of course the best is to also have a unit test proving the feature